src/firmware_main_loop.c

```diff
@@ -14,16 +14,25 @@
 
 #include "firmware_main_loop.h"
 
+#include "hww.h"
+#include "u2f.h"
 #include "ui/screen_process.h"
+#include "usb/usb.h"
 #include "usb/usb_processing.h"
 
 void firmware_main_loop(void)
 {
     while (1) {
         screen_process();
-        usb_processing_process(usb_processing_hww());
+        if (usb_is_enabled()) {
+            usb_processing_process(usb_processing_hww());
 #if APP_U2F == 1
-        usb_processing_process(usb_processing_u2f());
+            usb_processing_process(usb_processing_u2f());
```

Comment:
I don't think it's strictly necessary, but I feel that it's better to process by layers (so all the USB processing first, all application processing afterwards), so that events can propagate from the bottom layer in a single pass of the event loop.

Reply:
7471885